File: lab3/src/cpp/src/stack.cpp
#include "stack.hpp"

void Stack::push(std::string_view value) {
    m_list.push_head(value);
}

std::string Stack::pop() {
    if (m_list.getSize() == 0) {
        return {};
    }
    std::string value = m_list.get(0);
    m_list.del_head();
    return value;
}

std::string Stack::top() const {
    return m_list.get(0);
}

size_t Stack::size() const noexcept {
    return m_list.getSize();
}

bool Stack::empty() const noexcept {
    return m_list.getSize() == 0;
}

std::string Stack::print() const {
    if (empty()) {
        return "[] (top->bottom)";
    }
    return m_list.print() + " (top->bottom)";
}

File: lab3/src/go/internal/stack.go
package internal

type Stack struct {
	list *SinglyList
}

func NewStack() *Stack {
	return &Stack{
		list: NewSinglyList(),
	}
}

func (s *Stack) Push(value string) {
	s.list.PushHead(value)
}

func (s *Stack) Pop() string {
	if s.list.GetSize() == 0 {
		return ""
	}
	val := s.list.head.Data
	s.list.DelHead()
	return val
}

func (s *Stack) Top() string {
	if s.list.GetSize() == 0 {
		return ""
	}
	return s.list.head.Data
}

func (s *Stack) Size() int {
	return s.list.GetSize()
}

func (s *Stack) Empty() bool {
	return s.list.GetSize() == 0
}

func (s *Stack) Print() string {
	if s.Empty() {
		return "[] (top->bottom)"
	}
	return s.list.Print() + " (top->bottom)"
}